### ISSUE TITLE
[DevTools Bug]: TypeError: Cannot read properties of undefined (reading 'forEach')

### ISSUE THREAD

**Issue by nandkishorakangire** [facebook/react#27244]
### Website or app

the NextJS App in development phase

### Repro steps

On refreshing the web pages error being shown sometimes

<img width="722" alt="image" src="https://github.com/facebook/react/assets/56193250/bc3e26a3-243b-4e02-8d45-7c2f4a838323">

### How often does this bug happen?

Sometimes

### DevTools version (automated)

4.28.0 (7/10/2023)

### Error message (automated)

TypeError: Cannot read properties of undefined (reading 'forEach')

### Error call stack (automated)

```text
setup
chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/backendManager.js (4073:17)
welcome
chrome-extension://fmkadmapgofadopljbjfkapdkoienihi/build/backendManager.js (4057:3)
```

**Comment by tusharrana786** [comment]
this error occurs if you are trying to call forEach for an undefined value can you provide the source code or the snippet where you re getting this error

**Comment by Senseijacoby** [comment]
sometimes it's as easy as updating your system. if you are using styled components then see the updated documentation

**Comment by hoxyq** [comment]
Thanks for reporting this!

This seems like a really rare case, it fails [here](https://github.com/facebook/react/blob/main/packages/react-devtools-extensions/src/backendManager.js#L60-L62). If global hook is not registered yet, it probably means that `installHook` content script was not executed or has failed to execute for some reason.

We could add some explicit check before calling `hook.renderers.forEach`, contributions are welcomed.
But this won't fix the source of the problem, need to figure out why `__REACT_DEVTOOLS_GLOBAL_HOOK__` was not injected.

Any other logs / reproducible steps that can be shared?

**Comment by livwvil** [comment]
I faced with the same issue. My setup is:
```json
"react": "^18.2.0"
"react-dom": "^18.2.0"
"vite": "^4.3.9"
"@vitejs/plugin-react": "^3.1.0"
"@ch1ffa/vite-plugin-relay": "^2.3.0"
"react-relay": "^15.0.0",
"relay-runtime": "^15.0.0",
```
Sometimes when I change the source code the page reloads and this exception appears. It appears only after the first reload of the page automatically or manually.

**Comment by SFin12** [comment]
I'm having the same error and I get it randomly using a window.location.reload().

**Comment by daukantas** [comment]
Your github input has been violating my email !

Attention:  Please  make adjusted  changes to prevent your input entering this email site.


Warning! Please review “Scams and Identity Theft Attempts Targeting International Affiliates”, which can be found at https://www.hupd.harvard.edu/scams-targeting-international-affiliates
Scams and Identity Theft Attempts Targeting International Affiliates<https://www.hupd.harvard.edu/scams-targeting-international-affiliates>
Introduction
www.hupd.harvard.edu<http://www.hupd.harvard.edu/>
.

________________________________

**Comment by nandkishorakangire** [comment]
> Thanks for reporting this!
> 
> This seems like a really rare case, it fails [here](https://github.com/facebook/react/blob/main/packages/react-devtools-extensions/src/contentScripts/backendManager.js#L60-L62). If global hook is not registered yet, it probably means that `installHook` content script was not executed or has failed to execute for some reason.
> 
> We could add some explicit check before calling `hook.renderers.forEach`, contributions are welcomed. But this won't fix the source of the problem, need to figure out why `__REACT_DEVTOOLS_GLOBAL_HOOK__` was not injected.
> 
> Any other logs / reproducible steps that can be shared?

<img width="668" alt="image" src="https://github.com/facebook/react/assets/56193250/f5643e0a-7138-46b4-864f-3eac2be4b163">

Hi, @hoxyq Thanks for indetification of the issue. here's logs from browser's source for the errors shown in browser console. steps are as already mentioned, 'error being produced while page refreshed'. there was no issue occured from our code.

**Comment by bintus-ux** [comment]
I am also facing the same issue

**Comment by Shaker-Hamdi** [comment]
This issue is caused by the "react dev tool" extension on Chrome, uninstalling the extension solves this issue.

**Comment by livwvil** [comment]
> This issue is caused by the "react dev tool" extension on Chrome, uninstalling the extension solves this issue.

Oh, nice solution!

**Comment by hoxyq** [comment]
Hey, the browser extension v4.28.4 has been published today on Chrome's extensions store. This version includes various changes to the content scripts injection logic.

Can someone try updating the extension and check if this is still reproducible?

**Comment by livwvil** [comment]
> Hey, the browser extension v4.28.4 has been published today on Chrome's extensions store. This version includes various changes to the content scripts injection logic.
> 
> Can someone try updating the extension and check if this is still reproducible?

Looks like my case is solved. Thanks a lot!

**Comment by hoxyq** [comment]
Closing due to inactivity. Feel free to re-open if you have a reproducer with the latest version of the extension.
